File: remoteddns.php
<?php
// Variables

// Logfile to log output to
$logfile = '/var/log/nginx/ddns/remote.log'; //Make sure the path is writable

// The authentication key
define('KEY', 'Mjs74k9bC3ACBn5pKn9eCnPeBRZFrreY'); //Generate your own random key here

// SCRIPT
// Get external ip address
$ipAddress = $_SERVER['REMOTE_ADDR'];

// Check if domain is defined
if(empty($_GET['domain']))
{
    echo $ipAddress ; exit();
}

// The domain to edit
$fqdn = $_GET['domain'];

// Exit if the key does not match
if($_GET['key'] != KEY)
{
    echo "Error, wrong key."; exit();
}

// Include the required files from the API
require_once('lib/Transip/DomainService.php');

//set do_change to 0
$do_change = 0;
    
// Seperate subdomain and domain from fqdn
$subdomain = implode('.', explode('.', $fqdn, -2));
$domainname = array_slice(explode(".", $fqdn), -2)[0].".".array_slice(explode(".", $fqdn), -2)[1];

// Set record that has to be changed. 
$newValues = [
$subdomain => $ipAddress,
];

// Get the current DNS entries from TransIP
$dnsEntries = Transip_DomainService::getInfo($domainname)->dnsEntries;

// Loop through records
foreach($dnsEntries as $dnsEntry)
{
    // Check records for defined subdomain A record
    if (($dnsEntry->type == Transip_DnsEntry::TYPE_A) && ($dnsEntry->name == $subdomain))
    {
        // Check if the A record has to be updated
        if ($dnsEntry->content != $ipAddress)
        {
            $time = date('d/m/Y h:i:s a', time());
            file_put_contents($logfile, "". $time ." Updating old ip " . $dnsEntry->content . ", with: ". $ipAddress . " for: " . $subdomain . ".". $domainname . ".\r\n", FILE_APPEND);

            echo "Updating old ip " . $dnsEntry->content . ", with: ". $ipAddress . " for: " . $subdomain . ".". $domainname .PHP_EOL;
            $dnsEntry->content = $newValues[$dnsEntry->name];
            $do_change=1;
        }
        else
        {
            $time = date('d/m/Y h:i:s a', time());
            file_put_contents($logfile, "". $time ." No update required, current IP: " . $dnsEntry->content . " is unchanged for: " . $subdomain . ".". $domainname . ".\r\n", FILE_APPEND);
            
            echo "No update required, current IP: " . $dnsEntry->content . " is unchanged for: " . $subdomain . ".". $domainname .PHP_EOL;
        }
        break;
    }
    
}

// Update the record when nessecary
if ($do_change == 1 )
{
    try
    {
      // Commit the changes to the TransIP DNS servers
      Transip_DomainService::setDnsEntries($domainname, $dnsEntries);
      // Done
      echo 'DNS updated successfully!' .PHP_EOL;
    }
    catch(SoapFault $f)
    {
      // An error occured
      echo 'DNS not updated.' . $f->getMessage() .PHP_EOL;
    }
}
?>

File: src/remoteddns.py
#!/usr/bin/env python3
import os
from datetime import datetime

from flask import Flask, request
from transip import TransIP

app = Flask(__name__)

# Variables

# Logfile to log output to
logfile = '/var/log/ddns.log'

# The authentication key (set via REMOTEKEY environment variable)
KEY = os.environ.get('REMOTEKEY', 'Mjs74k9bC3ACBn5pKn9eCnPeBRZFrreY')  # Generate your own random key here


def log(message):
    """Append a timestamped message to the logfile."""
    time_str = datetime.now().strftime('%d/%m/%Y %I:%M:%S %p')
    with open(logfile, 'a') as f:
        f.write(f"{time_str} {message}\r\n")


def get_transip_client():
    """Return a configured TransIP client."""
    try:
        from lib import transip_settings
        login = transip_settings.LOGIN
        private_key = transip_settings.PRIVATE_KEY
        return TransIP(login=login, private_key=private_key)
    except (ImportError, AttributeError):
        pass

    login = os.environ.get('TRANSIP_LOGIN', '')
    private_key = os.environ.get('TRANSIP_PRIVATE_KEY', '')
    private_key_file = os.environ.get('TRANSIP_PRIVATE_KEY_FILE', '')

    if not login:
        raise RuntimeError('No TransIP login configured.')
    if private_key:
        return TransIP(login=login, private_key=private_key)
    if private_key_file:
        return TransIP(login=login, private_key_file=private_key_file)
    raise RuntimeError('No private key configured.')


@app.route('/', defaults={'path': ''})
@app.route('/<path:path>')
def remoteddns(path):
    # Get external ip address
    ip_address = request.remote_addr

    # Check for HTTP_X_FORWARDED_FOR header and use that if set
    x_forwarded_for = request.headers.get('X-Forwarded-For')
    if x_forwarded_for:
        ip_address = x_forwarded_for

    # Check if domain is defined
    fqdn = request.args.get('domain')
    if not fqdn:
        return ip_address

    # Exit if the key does not match
    if request.args.get('key') != KEY:
        return 'Error, wrong key.'

    try:
        client = get_transip_client()
    except RuntimeError as e:
        return str(e)

    do_change = False
    record_found = False

    # Separate subdomain and domain from fqdn
    parts = fqdn.split('.')
    subdomain = '.'.join(parts[:-2])
    domainname = '.'.join(parts[-2:])

    # Get the current DNS entries from TransIP
    domain = client.domains.get(domainname)
    dns_entries = domain.dns.list()

    result_lines = []

    # Loop through records
    for dns_entry in dns_entries:
        # Check records for defined subdomain A record
        if dns_entry.type == 'A' and dns_entry.name == subdomain:
            # Check if the A record has to be updated
            if dns_entry.content != ip_address:
                log(f"Updating old ip {dns_entry.content}, with: {ip_address} for: {subdomain}.{domainname}.")
                result_lines.append(
                    f"Updating old ip {dns_entry.content}, with: {ip_address} "
                    f"for: {subdomain}.{domainname}"
                )
                dns_entry.content = ip_address
                do_change = True
                record_found = True
            else:
                log(f"No update required, current IP: {dns_entry.content} is unchanged for: {subdomain}.{domainname}.")
                result_lines.append(
                    f"No update required, current IP: {dns_entry.content} "
                    f"is unchanged for: {subdomain}.{domainname}"
                )
                record_found = True
            break

    if not record_found:
        result_lines.append(f"Record: {subdomain}.{domainname} not found on TransIP Nameservers.")

    # Update the record when necessary
    if do_change:
        try:
            # Commit the changes to the TransIP DNS servers
            domain.dns.replace(dns_entries)
            # Done
            result_lines.append('DNS updated successfully!')
        except Exception as e:
            # An error occurred
            result_lines.append(f'DNS not updated. {str(e)}')

    return '\n'.join(result_lines)


if __name__ == '__main__':
    # Development server only. Use Gunicorn for production:
    # gunicorn -w 4 -b 0.0.0.0:80 remoteddns:app
    app.run(host='0.0.0.0', port=80)
